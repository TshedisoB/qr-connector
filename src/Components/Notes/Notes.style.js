import { StyleSheet } from "react-native";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
    padding: 16,
  },
  label: {
    fontSize: 15,
    marginBottom: 8,
  },
  input: {
    borderWidth: 1,
    borderColor: "#ccc",
    borderRadius: 4,
    padding: 8,
    marginBottom: 8,
  },
  buttonsContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    width: "60%",
  },
});

export { styles };
