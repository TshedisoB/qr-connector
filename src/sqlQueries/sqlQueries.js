export const createTable = (db) => {
  db.transaction((tx) => {
    tx.executeSql(`
      CREATE TABLE IF NOT EXISTS Contacts (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        email TEXT,
        name TEXT,
        profilePicLink TEXT,
        status TEXT,
        website TEXT,
        isFavorite BOOLEAN,
        notes TEXT
      )
    `);
  });
};

export const insertToDataBase = (db, data) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT id FROM Contacts WHERE email = ?;",
      [data.email],
      (_, resultSet) => {
        if (resultSet.rows.length === 0) {
          tx.executeSql(
            "INSERT INTO Contacts (email, name, profilePicLink, status, website, isFavorite, notes) VALUES (?, ?, ?, ?, ?, ?, ?)",
            [
              data.email,
              data.name,
              data.profilePicLink,
              data.status,
              data.website,
              data.isFavorite,
              data.notes,
            ]
          );
        }
      }
    );
  });
};

export const getAllContacts = (db, callback) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT * FROM Contacts ORDER BY isFavorite DESC",
      [],
      (_, resultSet) => {
        const contacts = resultSet.rows;
        const contactsArray = [];
        for (let i = 0; i < contacts.length; i++) {
          contactsArray.push(contacts.item(i));
        }
        callback(contactsArray);
      },
      () => {
        callback([]);
      }
    );
  });
};

export const updateFavorites = (db, email, isFavorite) => {
  db.transaction((tx) => {
    tx.executeSql("UPDATE Contacts SET isFavorite = ? WHERE email = ?;", [
      !isFavorite,
      email,
    ]);
  });
};

export const deleteContact = (db, email, callback) => {
  db.transaction((tx) => {
    tx.executeSql(
      "DELETE FROM Contacts WHERE email = ?;",
      [email],
      (_, resultSet) => {
        const rowsAffected = resultSet.rowsAffected;
        callback(rowsAffected > 0);
      },
      () => {
        callback(false);
      }
    );
  });
};

export const addOrUpdateNotes = (db, email, notes) => {
  db.transaction((tx) => {
    tx.executeSql(
      "SELECT * FROM Contacts WHERE email = ?;",
      [email],
      (_, resultSet) => {
        const existingContact = resultSet.rows.item(0);

        if (existingContact) {
          const updatedContact = { ...existingContact, notes };

          tx.executeSql("UPDATE Contacts SET notes = ? WHERE email = ?;", [
            updatedContact.notes,
            updatedContact.email,
          ]);
        }
      }
    );
  });
};
