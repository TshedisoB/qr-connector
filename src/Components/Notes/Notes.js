import React, { useState, useEffect } from "react";
import { useDispatch } from "react-redux";
import { View, Text, TextInput, Button } from "react-native";

import { addOrUpdateNotes } from "../../sqlQueries/sqlQueries";
import { setCurrentContact } from "../../Redux/actions";
import { styles } from "./Notes.style.js";
import { db } from "../ScanQr/ScanQr";

const Notes = ({ contact }) => {
  const [notes, setNotes] = useState(contact[0].notes);
  const [originalNotes, setOriginalNotes] = useState(contact[0].notes);
  const [isEditing, setIsEditing] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {
    const updatedContact = [{ ...contact[0], notes }];
    dispatch(setCurrentContact(updatedContact));
  }, [notes]);

  useEffect(() => {
    setNotes(contact[0].notes);
    setOriginalNotes(contact[0].notes);
  }, [contact]);

  useEffect(() => {
    if (!isEditing) {
      addOrUpdateNotes(db, contact[0].email, notes);
    }
  }, [isEditing, contact, notes]);

  const handleSaveNotes = () => {
    setIsEditing(false);
    setOriginalNotes(notes);
  };

  const handleCancelEdit = () => {
    setNotes(originalNotes);
    setIsEditing(false);
  };

  const handleNotesChange = (text) => {
    setNotes(text);
    setIsEditing(text !== originalNotes);
  };

  return (
    <View style={styles.container}>
      <Text style={styles.label}>Additional Contact Notes</Text>
      <TextInput
        style={styles.input}
        placeholder="You can make notes here..."
        value={notes}
        onChangeText={handleNotesChange}
        multiline
      />
      {isEditing && (
        <View style={styles.buttonsContainer}>
          <Button title="Cancel" onPress={handleCancelEdit} />
          <Button title="Save" onPress={handleSaveNotes} />
        </View>
      )}
    </View>
  );
};

export default Notes;
