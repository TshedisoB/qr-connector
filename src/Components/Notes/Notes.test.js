import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { createStore } from "redux";
import { Provider } from "react-redux";

import * as sqlQueries from "../../sqlQueries/sqlQueries";
import barCodeReducer from "../../Redux/reducers";
import Notes from "./Notes";

const store = createStore(barCodeReducer);

jest.mock("../../sqlQueries/sqlQueries", () => ({
  __esModule: true,
  addOrUpdateNotes: jest.fn(),
}));

describe("Notes component", () => {
  const contact = [
    {
      email: "tshediso.boshiana@umuzi.org",
      notes: "Initial notes",
    },
  ];

  it("should render correctly", () => {
    const { toJSON } = render(
      <Provider store={store}>
        <Notes contact={contact} />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("should display initial notes", () => {
    const { getByDisplayValue } = render(
      <Provider store={store}>
        <Notes contact={contact} />
      </Provider>
    );
    expect(getByDisplayValue("Initial notes")).toBeTruthy();
  });

  it("should update notes when edited", () => {
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <Notes contact={contact} />
      </Provider>
    );

    const notesInput = getByPlaceholderText("You can make notes here...");

    fireEvent.changeText(notesInput, "Updated notes");

    expect(notesInput.props.value).toBe("Updated notes");
  });

  it("should reset notes to initial value when Cancel button is pressed", () => {
    const { getByText, getByPlaceholderText } = render(
      <Provider store={store}>
        <Notes contact={contact} />
      </Provider>
    );

    const notes = "Updated notes for Tshediso";
    const inputText = getByPlaceholderText("You can make notes here...");

    fireEvent.changeText(inputText, notes);

    expect(inputText.props.value).toBe(notes);

    const cancelButton = getByText("Cancel");
    fireEvent.press(cancelButton);

    expect(inputText.props.value).toBe("Initial notes");
  });

  it("should call addOrUpdateNotes when Save button is pressed", () => {
    const { getByText, getByPlaceholderText } = render(
      <Provider store={store}>
        <Notes contact={contact} />
      </Provider>
    );

    const notes = "Updated notes for Tshediso";
    const inputText = getByPlaceholderText("You can make notes here...");

    fireEvent.changeText(inputText, notes);

    expect(inputText.props.value).toBe(notes);

    const saveButton = getByText("Save");
    fireEvent.press(saveButton);

    expect(sqlQueries.addOrUpdateNotes).toHaveBeenCalledWith(
      expect.any(Object),
      contact[0].email,
      notes
    );
  });
});
