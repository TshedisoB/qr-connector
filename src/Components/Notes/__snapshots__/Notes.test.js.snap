// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes component should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "padding": 16,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 8,
      }
    }
  >
    Additional Contact Notes
  </Text>
  <TextInput
    multiline={true}
    onChangeText={[Function]}
    placeholder="You can make notes here..."
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 4,
        "borderWidth": 1,
        "marginBottom": 8,
        "padding": 8,
      }
    }
    value="Initial notes"
  />
</View>
`;
